packages/bvaughn-architecture-demo/src/utils/suspense.ts

```diff
@@ -12,6 +14,27 @@ export function createWakeable<T>(): Wakeable<T> {
   let status: "unresolved" | "resolved" | "rejected" = "unresolved";
   let data: T | Error | null = null;
 
+  let callbacksRegisteredDuringCurrentFrame = 0;
+  let resetCounterTimer: NodeJS.Timeout | null = null;
+
+  // Guard against a case where promise resolution results in a new wakeable listener being added.
+  // That cause would result in an infinite loop.
+  // Note that our guard counter should be somewhat high to avoid false positives.
+  // It is a legitimate use-case to register handlers after a wakeable has been resolved or rejected.
+  const checkCircularThenableChain = () => {
+    if (++callbacksRegisteredDuringCurrentFrame > SYNCHRONOUS_CIRCULAR_THENABLE_CHECK_MAX_COUNT) {
+      throw Error("Circular thenable chain detected (infinite loop?)");
+    }
+
+    if (resetCounterTimer === null) {
+      resetCounterTimer = setTimeout(() => {
+        resetCounterTimer = null;
+
+        callbacksRegisteredDuringCurrentFrame = 0;
```

Comment:
Why reset the count? Shouldn't we just have one count per `wakeable`?
This reset might overlook infinite loops that are triggered asynchronously via non-microtasks?

Reply:
Perhaps it’s an unnecessary detail, but I was mostly trying to avoid potential false positives.